cmd/resource-quota-check/main.go

```diff
@@ -0,0 +1,141 @@
+	// Expects a comma separated list of values (i.e. "default,kube-system")
+	// Ignores namespaces from the blacklist if turned on. (On by default, checks all namespaces)
+	blacklistOnEnv         = os.Getenv("BLACKLIST")
+	blacklistNamespacesEnv = os.Getenv("BLACKLIST_NAMESPACES")
+	blacklistOn            = true
+	blacklistNamespaces    []string
+
+	// Looks at specified namespaces from the whitelist if turned on.
+	whitelistOnEnv         = os.Getenv("WHITELIST")
+	whitelistNamespacesEnv = os.Getenv("WHITELIST_NAMESPACES")
+	whitelistOn            bool
+	whitelistNamespaces    []string
+
+	// By default, ignore the "default" namespace.
+	defaultBlacklistNamespaces = []string{"default"}
+
+	// By default, look at "kube-system" and "kuberhealthy".
+	defaultWhitelistNamespaces = []string{"kube-system", "kuberhealthy"}
+
+	// Threshold for resource quota usage. (inclusive)
+	// If given 0.9 (or 90%), this check will alert when usage for memory or CPU is at least 90%.
+	thresholdEnv = os.Getenv("THRESHOLD")
+	threshold    float64
+
+	// Check time limit.
+	checkTimeLimitEnv = os.Getenv("CHECK_TIME_LIMIT")
+	checkTimeLimit    time.Duration
+
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	// Interrupt signal channels.
+	signalChan chan os.Signal
+	doneChan   chan bool
+
+	debugEnv = os.Getenv("DEBUG")
+	debug    bool
+
+	// K8s client used for the check.
+	client *kubernetes.Clientset
+
+	// checkErrors = make([]string, 0)
+)
+
+func init() {
+	// Parse incoming debug settings.
+	parseDebugSettings()
+
+	// Parse all incoming input environment variables and crash if an error occurs
+	// during parsing process.
+	parseInputValues()
+
+	// Allocate channels.
+	signalChan = make(chan os.Signal, 3)
+	doneChan = make(chan bool)
+}
+
+func main() {
+
+	ctx, ctxCancel = context.WithTimeout(context.Background(), time.Duration(time.Minute*5))
+
+	// Create a kubernetes client.
+	var err error
+	client, err = kubeClient.Create(kubeConfigFile)
+	if err != nil {
+		errorMessage := "failed to create a kubernetes client with error: " + err.Error()
+		kh.ReportFailure([]string{errorMessage})
+		return
+	}
+	log.Infoln("Kubernetes client created.")
+
+	// Catch panics.
+	var r interface{}
+	defer func() {
+		r = recover()
+		if r != nil {
+			log.Infoln("Recovered panic:", r)
+			kh.ReportFailure([]string{r.(string)})
+		}
+	}()
+
+	runResourceQuotaCheck()
+}
+
+func listenForInterrupts(ctx context.Context) {
+
+	// Relay incoming OS interrupt signals to the signalChan.
+	signal.Notify(signalChan, os.Interrupt, os.Kill, syscall.SIGTERM, syscall.SIGINT)
+	sig := <-signalChan // This is a blocking operation -- the routine will stop here until there is something sent down the channel.
+	log.Infoln("Received an interrupt signal from the signal channel.")
+	log.Debugln("Signal received was:", sig.String())
+
+	log.Debugln("Cancelling context.")
+	ctxCancel() // Causes all functions within the check to return without error and abort. NOT an error
```

Comment:
normally this is passed into the function, or you can run `ctx.Cancel()` I think.